dvc/repo/gc.py

```diff
@@ -41,12 +42,13 @@ def gc(
 
         used = NamedCache()
         for repo in all_repos + [self]:
+            has_scm = not isinstance(repo.scm, NoSCM)
             used.update(
                 repo.used_cache(
-                    all_branches=all_branches,
+                    all_branches=not all_branches and has_scm,
```

Comment:
I decided not to change the arguments name, as it makes sense to be more aggressive by setting the arguments here. and looks to be good name. The other idea I had was `scm_exclude=None` where we could pass *Collections* (set/tuple/list/set/dict, etc).

Reply:
@skshetry I don't think you need to pass `has_scm`, brancher is able to deal with no-scm case by itself.

Reply:
That does not seem to be the case. The `brancher` uses `SCM::resolve_rev()` which is not available on `NoSCM`.
```sh
$ dvc init --no-scm
$ python -c '
from dvc.repo import Repo
repo = Repo(".")
print(list(repo.brancher(all_branches=True, all_commits=True, all_tags=True)))
'
Traceback (most recent call last):
  File "<string>", line 4, in <module>
  File "/home/saugat/repos/iterative/dvc/dvc/repo/brancher.py", line 45, in brancher
    for sha, names in group_by(scm.resolve_rev, revs).items():
AttributeError: 'NoSCM' object has no attribute 'resolve_rev'
```

Reply:
@skshetry Oh, that is a very good catch! But the issue is that `scm.resolve_rev` gets resolved even for empty revs, hence why this problem raises. It needs to be fixed there, there is no need for `has_scm`, which is more of a workaround than a solution.

Reply:
I was confused with the usecase of `brancher`. From it's docstring:
> Generator that iterates over specified revisions.

So, I thought, it should not even work for `NoSCM`.

Reply:
@skshetry Got it. Yeah, we use NoSCM as a corner case, that doesn't have any branches, any commits, any tags, but does have the workspace 🙂 So most of the methods just don't return anything, which makes `brancher` work fine in that corner case of "no branches or revisions" 🙂